x-pack/plugin/ml-package-loader/src/main/java/org/elasticsearch/xpack/ml/packageloader/action/ModelImporter.java

```diff
@@ -119,11 +127,13 @@ public void doImport() throws URISyntaxException, IOException, ElasticsearchStat
             true
         );
 
-        executeRequestIfNotCancelled(PutTrainedModelDefinitionPartAction.INSTANCE, finalModelPartRequest);
+        executeRequestIfNotCancelled(PutTrainedModelDefinitionPartAction.INSTANCE, finalModelPartRequest, releasingListener);
+        logger.info("waiting for finish");
+        requestLimiter.acquire(MAX_IN_FLIGHT_REQUESTS); // cannot acquire until all inflight requests have completed
```

Comment:
Same here, we shouldn't block the thread to wait for the end of all downloads. A completion listener should be used to orchestrate the remaining work when the last download completed.